src/stcal/ramp_fitting/src/slope_fitter.c

```diff
@@ -0,0 +1,3561 @@
+    seg->end = end;
+    seg->length = end - start;
+    seg->flink = NULL;
+
+    /* Add segment to list as the tail */
+    if (NULL == segs->head) {
+        segs->head = seg;
+        segs->size = 1;
+    } else {
+        segs->tail->flink = seg;
+        segs->size++;
+    }
+    segs->tail = seg;
+
+    /* Is the new segment length the longest segment length? */
+    if (seg->length > segs->max_segment_length) {
+        segs->max_segment_length = seg->length;
+    }
+
+    return 0;
+}
+
+/*
+ * Build the optional results class to be returned from ramp fitting.
+ */
+static PyObject *
+build_opt_res(
+        struct ramp_data * rd) /* The ramp fitting data */
+{
+    struct opt_res_product opt_res = {0};
+    PyObject * opt_res_info = Py_None;
+
+    /* Make PyObjectArray's */
+    if (create_opt_res(&opt_res, rd)) {
+        return Py_None;
+    }
+
+    /* Copy data from rd->segs to these arrays */
+    save_opt_res(&opt_res, rd);
+
+    /* Package arrays into output tuple */
+    opt_res_info = Py_BuildValue("NNNNNNNNN",
+        opt_res.slope, opt_res.sigslope, opt_res.var_p, opt_res.var_r,
+        opt_res.yint, opt_res.sigyint, opt_res.pedestal, opt_res.weights,
+        opt_res.cr_mag);
+
+    return opt_res_info;
+}
+
+/*
+ * Clean up all allocated memory for a pixel ramp, except the allocated memory
+ * for the data structure itself.
+ */
+static void
+clean_pixel_ramp(
+        struct pixel_ramp * pr)  /* Ramp fitting data for a pixel. */
+{
+    if (NULL == pr) {
+        return;  /* Nothing to do */
+    }
+
+    /* Free all internal arrays */
+    SET_FREE(pr->data);
+    SET_FREE(pr->groupdq);
+    SET_FREE(pr->rateints);
+    SET_FREE(pr->stats);
+    SET_FREE(pr->is_zframe);
+    SET_FREE(pr->is_0th);
+
+    /* Clean up the allocated memory for the linked lists. */
+    FREE_SEGS_LIST(pr->nints, pr->segs);
+}
+
+/* Cleans up the ramp data structure */
+static void
+clean_ramp_data(
+        struct ramp_data * rd) /* The ramp fitting data structure */
+{
+    npy_intp idx;
+    struct simple_ll_node * current;
+    struct simple_ll_node * next;
+
+    if (NULL == rd->segs) {
+        return; /* Nothing to do. */
+    }
+
+    /*
+     *  For each pixel, check to see if there is any allocated
+     *  memory for the linked list of ramp segments and free them.
+     */
+    for (idx=0; idx < rd->cube_sz; ++idx) {
+        current = rd->segs[idx];
+        if (current) {
+            next = current->flink;
+            SET_FREE(current);
+            current = next;
+        }
+    }
+    SET_FREE(rd->segs);
+}
+
+/*
+ * Cleans up the rate producte data structure.
+ */
+static void
+clean_rate_product(
+        struct rate_product * rate_prod)  /* Rate product data structure */
+{
+    /* Free all arrays */
+    PyDataMem_FREE(rate_prod->slope);
+    PyDataMem_FREE(rate_prod->dq);
+    PyDataMem_FREE(rate_prod->var_poisson);
+    PyDataMem_FREE(rate_prod->var_rnoise);
+    PyDataMem_FREE(rate_prod->var_err);
```

Comment:
Maybe I already asked this. If so, please ignore. Why `PyDataMem_FREE` and not `Py_DECREF`? I am not familiar with `PyDataMem_FREE` and it is poorly documented. Can you  justify this?